### PR TITLE
[SPARK-42569][CONNECT] Throw exceptions for unsupported session API

### PR DESCRIPTION
**Pull request by amaliujia**
### What changes were proposed in this pull request?

Throw exceptions for unsupported session API:
1. newSession
2. getActiveSession
3. getDefaultSession
4. active

### Why are the changes needed?

Better indicate an API is not supported to users.

### Does this PR introduce _any_ user-facing change?

NO

### How was this patch tested?

N/A

### CODE REVIEW

**Comment by amaliujia** [comment]
@hvanhovell

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala [hunk 1]
```diff
@@ -190,6 +190,22 @@ class SparkSession(
     range(start, end, step, Option(numPartitions))
   }
 
+  def newSession(): SparkSession = {
+    throw new UnsupportedOperationException("newSession is not supported")
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala:194]: Perhaps we should have a special exception for this, so it is easier to catch these compatibility issues for users. We can also use them to provide migration hints. Finally, there is also a difference between the methods we do want to support, and the ones we don't.

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala [hunk 2]
```diff
@@ -190,6 +190,22 @@ class SparkSession(
...
+  }
+
+  def getActiveSession: Option[SparkSession] = {
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala:197]: These three a part of the object and not the class itself right?

**Comment by hvanhovell** [comment]
@amaliujia can you please update the compatibility test for these?

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala (hunk 1 above)
Comment (amaliujia) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala:194]: This is to match Python client: https://github.com/apache/spark/blob/5243d0be2c15e3af36e981a9487ea600ab86a808/python/pyspark/sql/connect/session.py#L422


I see the difference between `want to support but not yet` and `won't support`. Right now it is not differentiated in both Scala client and Python client. We probably need a pass to change both together if we want to do so.

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala (hunk 2 above)
Comment (amaliujia) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala]: right. updated

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CompatibilitySuite.scala [hunk 3]
```diff
@@ -194,10 +196,6 @@ class CompatibilitySuite extends ConnectFunSuite {
       ProblemFilters.exclude[Problem]("org.apache.spark.sql.SparkSession.this"),
       ProblemFilters.exclude[Problem]("org.apache.spark.sql.SparkSession.setActiveSession"),
       ProblemFilters.exclude[Problem]("org.apache.spark.sql.SparkSession.clearActiveSession"),
-      ProblemFilters.exclude[Problem]("org.apache.spark.sql.SparkSession.setDefaultSession"),
-      ProblemFilters.exclude[Problem]("org.apache.spark.sql.SparkSession.clearDefaultSession"),
-      ProblemFilters.exclude[Problem]("org.apache.spark.sql.SparkSession.getActiveSession"),
-      ProblemFilters.exclude[Problem]("org.apache.spark.sql.SparkSession.getDefaultSession"),
```
Comment (amaliujia) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CompatibilitySuite.scala:200]: These are repeating entries so can be safely deleted.

**Comment by amaliujia** [comment]
@hvanhovell Compatibility tests updated

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala (hunk 1 above)
Comment (hvanhovell) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala:223]: For the record I think we should implement this particular method.

Regarding the errors in general. I don't think we spend a great deal of time on what we did on the python side. I am fine with coordinating efforts. Can you suggest something when you have cycles? Long term I want the never supported errors to point to something that offers some explanation as to why, and I want the will support someday errors to point to a JIRA. cc @grundprinzip

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CompatibilitySuite.scala [hunk 4]
```diff
@@ -152,6 +153,8 @@ class CompatibilitySuite extends ConnectFunSuite {
       ProblemFilters.exclude[Problem]("org.apache.spark.sql.SparkSession.getActiveSession"),
       ProblemFilters.exclude[Problem]("org.apache.spark.sql.SparkSession.clearDefaultSession"),
       ProblemFilters.exclude[Problem]("org.apache.spark.sql.SparkSession.setDefaultSession"),
+      ProblemFilters.exclude[Problem]("org.apache.spark.sql.SparkSession.newSession"),
+      ProblemFilters.exclude[Problem]("org.apache.spark.sql.SparkSession.implicits"),
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CompatibilitySuite.scala:157]: Why does this one need to be added? I added `SQLImplicits` yesterday. Can you update your branch?

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CompatibilitySuite.scala [hunk 5]
```diff
@@ -174,10 +177,6 @@ class CompatibilitySuite extends ConnectFunSuite {
       ProblemFilters.exclude[Problem]("org.apache.spark.sql.SparkSession.this"),
       ProblemFilters.exclude[Problem]("org.apache.spark.sql.SparkSession.setActiveSession"),
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CompatibilitySuite.scala:178]: Also do these: `setActiveSession` & `clearActiveSession`?

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala (hunk 1 above)
Comment (amaliujia) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala:223]: I can follow up on this.

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CompatibilitySuite.scala (hunk 5 above)
Comment (amaliujia) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CompatibilitySuite.scala]: I guess someone else did a batch addition which added these two API?

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CompatibilitySuite.scala (hunk 4 above)
Comment (amaliujia) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CompatibilitySuite.scala:157]: maybe that was forgotten to be added?

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CompatibilitySuite.scala (hunk 5 above)
Comment (amaliujia) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CompatibilitySuite.scala]: I see. Done

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CompatibilitySuite.scala (hunk 4 above)
Comment (amaliujia) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CompatibilitySuite.scala:157]: Ok I removed it..

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CompatibilitySuite.scala [hunk 6]
```diff
@@ -129,6 +129,7 @@ class CompatibilitySuite extends ConnectFunSuite {
       ProblemFilters.exclude[Problem]("org.apache.spark.sql.Dataset.javaRDD"),
       ProblemFilters.exclude[Problem]("org.apache.spark.sql.Dataset.writeStream"),
       ProblemFilters.exclude[Problem]("org.apache.spark.sql.Dataset.this"),
+      ProblemFilters.exclude[Problem]("org.apache.spark.sql.Dataset.sparkSession"),
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CompatibilitySuite.scala:132]: Why are you adding this one? It used to work? What changed?

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CompatibilitySuite.scala [hunk 7]
```diff
@@ -152,6 +153,7 @@ class CompatibilitySuite extends ConnectFunSuite {
       ProblemFilters.exclude[Problem]("org.apache.spark.sql.SparkSession.getActiveSession"),
       ProblemFilters.exclude[Problem]("org.apache.spark.sql.SparkSession.clearDefaultSession"),
       ProblemFilters.exclude[Problem]("org.apache.spark.sql.SparkSession.setDefaultSession"),
+      ProblemFilters.exclude[Problem]("org.apache.spark.sql.SparkSession.newSession"),
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CompatibilitySuite.scala:156]: Same question?

**Comment by amaliujia** [comment]
@hvanhovell I should make the Compatibility list correct now.

**Review by hvanhovell — APPROVED** [review]
LGTM

**Comment by hvanhovell** [comment]
Merging

**Pull request closed without merging**